### PR TITLE
IServiceBroker/MessagePack/OOP fix roll-up

### LINKED ISSUE #48175: Report NFW when OCE is thrown by brokered service without cancellationToken signaled
Tracking TODO in BrokeredServiceBase.cs

### PR DESCRIPTION
**Pull request by sharwell**
Supersedes #47870, #48179, #48187, #48188, #48195

Closes #48175

### CODE REVIEW

File: src/Workspaces/Remote/Core/RemoteCallback.cs [hunk 1]
```diff
@@ -118,20 +108,24 @@ private bool ReportUnexpectedException(Exception exception, CancellationToken ca
...
             return FatalError.ReportWithoutCrashAndPropagate(exception);
         }
 
-        private static Exception OnUnexpectedException(CancellationToken cancellationToken)
+        private static Exception OnUnexpectedException(Exception exception, CancellationToken cancellationToken)
         {
             cancellationToken.ThrowIfCancellationRequested();
 
-            // If this is hit the cancellation token passed to the service implementation did not use the correct token.
+            if (exception is ConnectionLostException)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/RemoteCallback.cs:122]: this is probably worth documenting why we're doing it this way.  doesn't need to be in this PR though.

File: src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs [hunk 2]
```diff
@@ -113,20 +118,26 @@ public static ValueTask<ImmutableArray<(Checksum, object)>> ReadDataAsync(PipeRe
...
                 cancellationToken.ThrowIfCancellationRequested();
 
                 throw exception ?? ExceptionUtilities.Unreachable;
             }
+            finally
+            {
+                // Make sure to complete the copy and pipes before returning, otherwise the caller could complete the
+                // reader and/or writer while they are still in use.
+                await copyTask.ConfigureAwait(false);
+            }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs:140]: so many finallies :)

File: src/Workspaces/Remote/Core/ServiceDescriptor.cs [hunk 3]
```diff
@@ -78,6 +79,26 @@ protected override JsonRpcConnection CreateConnection(JsonRpc jsonRpc)
...
 
+        public override ServiceRpcDescriptor WithMultiplexingStream(MultiplexingStream? multiplexingStream)
+        {
+            var baseResult = base.WithMultiplexingStream(multiplexingStream);
+            if (baseResult is ServiceDescriptor)
+                return baseResult;
+
+            // work around incorrect implementation in 16.8 Preview 2
+            if (MultiplexingStream == multiplexingStream)
+                return this;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/ServiceDescriptor.cs:90]: possibly to doc/explain/link this to something clearer?

Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/Core/ServiceDescriptor.cs:90]: actually, doc the whole thing tbh, i don't undestand any of it :)

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Workspaces/Remote/Core/ServiceDescriptor.cs (hunk 3 above)
Comment (sharwell) [reply src/Workspaces/Remote/Core/ServiceDescriptor.cs:90]: See https://github.com/dotnet/roslyn/pull/47870#discussion_r496365307

Pull request #48153 will revert this as soon as roslyn-integration-CI is updated to 16.8 Preview 3 or newer.

File: src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs (hunk 2 above)
Comment (sharwell) [reply src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs:140]: so many

File: src/Workspaces/Remote/Core/RemoteCallback.cs (hunk 1 above)
Comment (sharwell) [reply src/Workspaces/Remote/Core/RemoteCallback.cs:122]: It's documented a few lines above, and still under investigation.

https://github.com/dotnet/roslyn/blob/4123b1c06ee17b42644931995ab00b15907cd8c6/src/Workspaces/Remote/Core/RemoteCallback.cs#L103-L112

**Pull request merged**